cmd/panfs-helpers.go

```diff
@@ -14,3 +63,169 @@ func dotS3PrefixCheck(objects ...string) error {
 	}
 	return nil
 }
+
+// Publish source path to destination path, creates all the
+// missing parents if they don't exist and change the rights.
+func panfsPublishFile(ctx context.Context, sourcePath, destPath string, objMode, dirMode os.FileMode, ownerID, groupID int) error {
+	if err := checkPathLength(sourcePath); err != nil {
+		logger.LogIf(ctx, err)
+		return err
+	}
+	if err := checkPathLength(destPath); err != nil {
+		logger.LogIf(ctx, err)
+		return err
+	}
+
+	if err := panPublishFileAll(sourcePath, destPath, objMode, dirMode, ownerID, groupID); err != nil {
+		logger.LogIf(ctx, err)
+		return err
+	}
+
+	return nil
+}
+
+func panPublishFileAll(srcFilePath, dstFilePath string, objMode, dirMode os.FileMode, ownerID, groupID int) (err error) {
+	if srcFilePath == "" || dstFilePath == "" {
+		return errInvalidArgument
+	}
+
+	if err = checkPathLength(srcFilePath); err != nil {
+		return err
+	}
+	if err = checkPathLength(dstFilePath); err != nil {
+		return err
+	}
+
+	if err = panReliablePublishFile(srcFilePath, dstFilePath, objMode, dirMode, ownerID, groupID); err != nil {
+		switch {
+		case isSysErrNotDir(err) && !osIsNotExist(err):
+			// Windows can have both isSysErrNotDir(err) and osIsNotExist(err) returning
+			// true if the source file path contains an non-existent directory. In that case,
+			// we want to return errFileNotFound instead, which will honored in subsequent
+			// switch cases
+			return errFileAccessDenied
+		case isSysErrPathNotFound(err):
+			// This is a special case should be handled only for
+			// windows, because windows API does not return "not a
+			// directory" error message. Handle this specifically here.
+			return errFileAccessDenied
+		case isSysErrCrossDevice(err):
+			return fmt.Errorf("%w (%s)->(%s)", errCrossDeviceLink, srcFilePath, dstFilePath)
+		case osIsNotExist(err):
+			return errFileNotFound
+		case osIsExist(err):
+			// This is returned only when destination is a directory and we
+			// are attempting a rename from file to directory.
+			return errIsNotRegular
+		default:
+			return err
+		}
+	}
+	return nil
+}
+
+// Reliably retries os.RenameAll if for some reason os.RenameAll returns
+// syscall.ENOENT (parent does not exist).
+func panReliablePublishFile(srcFilePath, dstFilePath string, objMode, dirMode os.FileMode, ownerID, groupID int) (err error) {
+	if err = panReliableMkdirAll(path.Dir(dstFilePath), dirMode, ownerID, groupID); err != nil {
+		return err
+	}
+
+	if err = os.Chmod(srcFilePath, objMode); err != nil {
+		return err
+	}
+	if err = os.Chown(srcFilePath, ownerID, groupID); err != nil {
+		return err
+	}
+	i := 0
+	for {
+		// After a successful parent directory create attempt a renameAll.
+		if err = PanRenameFile(srcFilePath, dstFilePath); err != nil {
+			// Retry only for the first retryable error.
+			if osIsNotExist(err) && i == 0 {
```

Comment:
to be honest, I do not understand the value of this retry logic in original minio implementation. First, I do not understand why "does not exist" is a "retryable" error. I can understand random EIO or EITNR, but ENOENT does not look as something may magically appear after retry. So it seems the condition should be ```!osIsNotExists(err) && i == 0```.
Even if we fix this condition, it is still not clear whether chown/chmod should be under this retry loop or not...